src/sources/util/unix_datagram.rs

```diff
@@ -26,16 +27,36 @@ use crate::{
 /// syslog source).
 pub fn build_unix_datagram_source(
     listen_path: PathBuf,
+    socket_file_mode: Option<u32>,
     max_length: usize,
     decoder: Decoder,
     handle_events: impl Fn(&mut [Event], Option<Bytes>) + Clone + Send + Sync + 'static,
     shutdown: ShutdownSignal,
     out: SourceSender,
-) -> Source {
-    Box::pin(async move {
-        let socket = UnixDatagram::bind(&listen_path).expect("Failed to bind to datagram socket");
-        info!(message = "Listening.", path = ?listen_path, r#type = "unix_datagram");
+) -> crate::Result<Source> {
+    let socket = UnixDatagram::bind(&listen_path).expect("Failed to bind to datagram socket");
+    info!(message = "Listening.", path = ?listen_path, r#type = "unix_datagram");
 
+    if let Some(socket_permission) = socket_file_mode {
+        match fs::set_permissions(&listen_path, fs::Permissions::from_mode(socket_permission)) {
```

Comment:
Would it make sense to [set `umask`](https://docs.rs/libc/latest/libc/fn.umask.html) around this call (and for `unix_stream`) to reduce the permissions race?
```suggestion
    let old_umask = socket_file_mode.map(|_| unsafe { libc::umask(0o777) });
    let socket = UnixDatagram::bind(&listen_path).expect("Failed to bind to datagram socket");
    info!(message = "Listening.", path = ?listen_path, r#type = "unix_datagram");

    if let Some(socket_permission) = socket_file_mode {
        unsafe { libc::umask(old_umask.unwrap()) };
        match fs::set_permissions(&listen_path, fs::Permissions::from_mode(socket_permission)) {
```

Reply:
Hum, how is it going to reduce the permissions race? :thinking: